### DISCUSSION TITLE
Experiencing unnecessary recompiles of bindgen

### DISCUSSION THREAD

**Discussion by jfly** [ipetkov/crane#518] [General]
I'm writing a simple flake using crane. I've diverged only very slightly from `nix flake init -t github:ipetkov/crane#quick-start-simple`, I've just added `v4l2-sys-mit` to my Cargo.toml (it's not particularly important, other than that it uses bindgen) and `rustPlatform.bindgenHook` to my derivation's `nativeBuildInputs`.

I've noticed that I'm not benefiting from caching of `cargoArtifacts` as much as I'd like. I've put together a simple repro on https://github.com/jfly/crane-bindgen/tree/repro.

I dug into this, and the problem ultimately comes down to: 

1. Each derivation gets a unique [`-frandom-seed=...`](https://github.com/NixOS/nixpkgs/blob/3a73796bf2edb1dc026257da827678117ee7af57/pkgs/build-support/setup-hooks/reproducible-builds.sh#L5) value in `NIX_CFLAGS_COMPILE`.
2. `rustPlatform.bindgenHook` then [interpolates that `NIX_CFLAGS_COMPILE` flag into `BINDGEN_EXTRA_CLANG_ARGS`](https://github.com/NixOS/nixpkgs/blob/3a73796bf2edb1dc026257da827678117ee7af57/pkgs/build-support/rust/hooks/rust-bindgen-hook.sh#L9).
3. rust-bindgen's build.rs [forces a rebuild if it sees that the `BINDGEN_EXTRA_CLANG_ARGS` env var changes](https://github.com/rust-lang/rust-bindgen/blob/v0.69.4/bindgen/build.rs#L20).

See my commit message here https://github.com/jfly/crane-bindgen/commit/4d12415ee8dcb05ba13caa3f86125435a3b26e1e for an explanation of how I determined this.

I am able to work around this with [a nasty hack to force `-frandom-seed` to the same value for both derivations](https://github.com/jfly/crane-bindgen/commit/de46de2a4a1e5e54f381c234da6d14753721496e) (there might be a better way of accomplishing this in the nix/nixpkgs ecosystem, I didn't dig too far).

Do you have any thoughts on a better fix for this? Perhaps crane's perspective is that this `cargo:rerun-if-env-changed=BINDGEN_EXTRA_CLANG_ARGS` is just something weird that bindgen does that isn't necessary in the nix ecosystem (where anything that would change this `BINDGEN_EXTRA_CLANG_ARGS` env var would trigger a rebuild from scratch anyways)?

**Comment by dpc** [comment]
https://github.com/ipetkov/crane/issues/370

```
[profile.dev.build-override]
debug = false
```

maybe?

**Comment by ipetkov** [comment]
Hi @jfly thanks for the detailed writeup!

One thing I noticed is that the [current version of `reproducible-builds.sh` in `nixpkgs-unstable`](https://github.com/NixOS/nixpkgs/blob/c0b7a892fb042ede583bdaecbbdc804acb85eabe/pkgs/build-support/setup-hooks/reproducible-builds.sh#L6) allows for explicitly setting this seed via `NIX_OUTPATH_USED_AS_RANDOM_SEED` (e.g. set it to any arbitrary 10 character string). I think setting this to the same value for both derivations should avoid the rebuild!

I'll document this for anyone else who runs into it in the future so thanks for mentioning it!

**Reply by ipetkov** [reply]
This will be documented in https://github.com/ipetkov/crane/pull/519 !
